Fix18756

bendk opened the pull request:
Janet gives this the thumbs up.  This is ready to merge, assuming the peer-review is okay.

The main point of all of this is to try to make device database handling work better as devices pass between miro versions.  This meant making it so that the current code outputs things in a way that previous versions can deal with.

Also it meant trying to change the metadata/metadata_status schemas in a way so that future versions can change the schema and still work with this version.  I'm thinking the process would go something like this:
- Suppose Miro 6.0 adds a new metadata processor called "foo"
- we add a foo_status column and set it to noneOk=True.  I think this means that older versions can still read/write to the database and when it writes to the database the column is set to NULL.
- if NULL isn't okay, we can add code that runs an upgrade by selecting all the rows where foo_status is NULL.
- We change the value that gets set to finished_status to 2.  This way if miro 6.0 sets finished_status, then Miro 5.0 will know that it doesn't need to run any metadata processors either.  But if Miro 5.0 sets it, then Miro 6.0 will still get a chance to run the new processor.

Commit:
bz18756 part 1: switch cover_art_path and screenshot_path back to old names

We shouldn't change the attribute names if we want device items to work when
the device is plugged into an old version.

Commit:
bz18756 part 2: fixes for devicedatabaseupgrade

Check for items that aren't using the metadata system each time we open a
device instead of only when we create the SQLite database.  This works better
when users go back and forth between versions.

We set some values for mdp_state, title_tag, and metadata_version, for
DeviceItems so that they're present for older miro versions.  The values we
set ensure that old versions won't run mutagen or MDP on the items, but should
see the metadata when never versions run mutagen/MDP.

Commit:
Make dialog signals raise errors in the unittests

This was helpful to help debuging things as I wrote the next for changesets

Commit:
bz18756 part 3: remove STATUS_SKIPPED_FROM_PREF

Instead, track the file_type in metadata_status.  It seems good to keep the
possible values of the status columns small.  It's also semi-likely that we'll
want to use file_type for future processors.

Commit:
bz18756 part 4: ditch current_processor

current_processor won't work as we add new processor types in newer versions.
Replaced it with a system that hopefully should work between versions.

Commit:
bz18756: Device importing improvements

Refactored the code that deals with failures when converting old databases.
Now it handles it on a row-by-row basis.  If one conversion fails, we insert
values as if the file was just added to the system and continue with the rest
of the items.  Changed _do_import_old_items() into a function object to try to
make it at least semi-understandable.

Added code to deal with databases with older versions by resetting them.  This
is a fairy heavy handed approach, but it's simple and only affects users who
have been running the nightlies.

Changed LiveStorage._get_version() to be a public method.  Made
reset_database() optionally initialize the schema as well.

willkg commented on tv/lib/devicedatabaseupgrade.py:
```diff
@@ -57,182 +61,197 @@ def import_from_json(live_storage, json_db, mount):
+        # track the next id that we should create in the database
+        self.id_counter = itertools.count(databaseupgrade.get_next_id(cursor))
+
+    def invoke(self):
```
You could put the invoke() code in **call** and thus make the class instance callable.

bendk replied:
Yeah, I went back and forth about that.  I went with a method name because writing the double parens in _DoImportOldItems()() seemed weird.

bendk replied:
Sweet to get some feedback from outside the paid dev team!  Thanks for the eyes.

Ben

On Thu, Feb 23, 2012 at 2:09 PM, Will Kahn-Greene
reply@reply.github.com
wrote:

> > -        cursor.execute("SELECT path FROM metadata_status")
> > -        self.paths_in_metadata_table = set(row[0] for row in cursor)
> >   +
> > -        # get info about each item on the device
> > -        self.device_items = []
> > -        for file_type in (u'audio', u'video', u'other'):
> > -            if file_type not in json_db:
> > -                continue
> > -            for path, data in json_db[file_type].iteritems():
> > -                if path not in self.paths_in_metadata_table:
> > -                    self.device_items.append((file_type, path, data))
> >   +
> > -        # track the next id that we should create in the database
> > -        self.id_counter = itertools.count(databaseupgrade.get_next_id(cursor))
> >   +
> > -    def invoke(self):
> 
> You could put the invoke() code in **call** and thus make the class instance callable.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/pculture/miro/pull/215/files#r481780

bendk replied:
After thinking about this some more, I think I'll just put the code in the constructor and rename the class name to _do_import_old_items, so running the code will be just like making a function call.

willkg commented on tv/lib/devicedatabaseupgrade.py:
```diff
@@ -57,182 +61,197 @@ def import_from_json(live_storage, json_db, mount):
+            finished_status = 1
+
+        if self.net_lookup_enabled:
+            echonest_status = 'N' # STATUS_NOT_RUN
```
You're better off putting these values into "constants" somewhere.

willkg commented on tv/lib/devices.py:
```diff
@@ -1338,11 +1338,20 @@ def load_sqlite_database(mount, json_db, device_size, countdown=0):
+        # upgrading are ones that have gone through the nightlys.  Just
+        # re-create the database in this case
+        logging.warn("Reseting device database: %r", mount)
+        live_storage.reset_database(init_db=True)
```
I'm not sure I understand the implications of this. If the database on the device isn't the right version, then will Miro treat the device as if it were new, go through the device from scratch and rebuild the database on the device? And that could potentially happen every time I upgrade Miro? That's pretty intense, if true.

bendk replied:
Yup that's exactly the implication.  I thought this code would be okay because 4.0.x didn't create databases, so I figured the only database where this would make a differences was ones created with the nightlies.  I just realize that we need to deal with versions after 5.0, and this code will definitely not do the right thing there.

I think I will change != to <, so that we won't erase databases created by newer versions of miro.  At that point it will be pretty ugly, but hopefully good enough to push out the door and we can rethink things for the next version.

willkg commented on tv/lib/item.py:
```diff
@@ -2427,6 +2427,11 @@ def __init__(self, **kwargs):
         self._fix_paths_from_database(kwargs)
+        # set values for attributes used in pre-5.0 databases.
+        self.metadata_version = 5 # version used in 4.0.x
+        self.mdp_state = 1 # RAN stat
```
This is another magic value that should be put in a "constant" somewhere.

Commit:
Fixes based on willkg's comments.

Use variables for constants rather than just setting the values.  I realized
that devicedatabaseupgrade is not like databaseupgrade in that we can safely
import other modules in that code and actually use the value.

Fix for our code that resets the database if the version is different than
what we expect.

Commit:
One more fix based on will's comments.

Renamed _DoImportOldItems and changed the way we invoke it so that it looks
like a function call.

paulswartz commented on tv/lib/devicedatabaseupgrade.py:
```diff
@@ -57,182 +61,197 @@ def import_from_json(live_storage, json_db, mount):
-        live_storage.cursor.execute("COMMIT TRANSACTION")
 
-def _do_import(cursor, json_db, mount):
+class _DoImportOldItems(object):
```
Why the CamelCase? Alternatively, why the strange class name?

bendk replied:
I think this note got messed up somehow, but it's funny that you and will pick apart different aspects of my function object :)  As of the latest commit the class is called _do_import_old_items()

paulswartz replied:
Yeah, it got moved. I just think that it should either be CamelCase (without the _ prefix) or _under_scored, but not both.

paulswartz commented on tv/lib/devicedatabaseupgrade.py:
```diff
@@ -57,182 +62,196 @@ def import_from_json(live_storage, json_db, mount):
-    for file_type, path, old_item in device_items:
+    def __init__(self, cursor, json_db, mount):
+        self.cover_art_dir = os.path.join(mount, '.miro', 'cover-art')
+        self.net_lookup_enabled = app.config.get(prefs.NET_LOOKUP_BY_DEFAULT)
```
This, `column_map`, and `insert_columns` could probably be class attributes.

paulswartz commented on tv/lib/item.py:
```diff
@@ -2458,9 +2467,9 @@ def __init__(self, **kwargs):
-        """Make screenshot_path and cover_art_path the correct type.
+        """Make screenshot and cover_art the correct type.
         """
-        for key in ('screenshot_path', 'cover_art_path'):
```
`title_tag` -> `title`?

Commit:
Cleanup _do_import_device_items

Took paul's suggestion of moving some attributes to the class-level instead of
setting them up in __init__()

Tried to break up the rest of the code in __init__()

Commit:
bz18808: using wrong parameter name

paulswartz commented on tv/lib/devicedatabaseupgrade.py:
```diff
@@ -113,126 +102,177 @@ def _do_import(cursor, json_db, mount):
+        # versions from running the movie data program on them.  This seems
+        # the safest option and old versions should still pick up new metadata
+        # when newer versions run MDP.
+        old_item['mdp_state'] = 1
```
Can this explicitly say MDP_STATUS_RAN since we're defining it?

bendk replied:
oops, I'll fix that.

paulswartz commented on tv/lib/devicedatabaseupgrade.py:
```diff
@@ -113,126 +102,177 @@ def _do_import(cursor, json_db, mount):
-                             OLD_ITEM_PRIORITY))
-        status_id = next_id
-        next_id += 1
+            finished_status = 1
```
What do 0 and 1 mean here?

bendk replied:
true/false.  It's a boolean field

paulswartz replied:
Can it use True/False then? Or something like (moviedata_status != 'N')?

bendk replied:
I was worried if sqlite would convert the values correctly, but I just checked and it does.

paulswartz commented on tv/lib/devices.py:
```diff
@@ -1338,11 +1338,24 @@ def load_sqlite_database(mount, json_db, device_size, countdown=0):
+        # FIXME: Need to write real upgrade code for post-5.0
+        logging.warn("Reseting device database: %r", mount)
+        live_storage.reset_database(init_schema=True)
+
```
Maybe log a warning for the else case too?

bendk replied:
Good point.  The else case is greater than right?  I think if DB_VERSION == get_version() we don't need to log anything.

paulswartz replied:
Ah yes. No need to log for ==.

paulswartz commented:
A couple documentation things, and a rebase (can't be automatically merged) and this should be all set.

Commit:
Use MDP_STATE_RAN rather than a hard-coded number

Commit:
Couple more tweaks from paul's suggestions.

Use True/False instead of 1/0 for boolean values.  Sqlite can convert them
fine.

Log a warning if we open a device sqlite database from a newer version of
miro.

bendk commented:
okay rebase is done and I believe all the changes that paul asked for were made.
